File: ts-reaktive-marshal/src/main/java/com/tradeshift/reaktive/xml/QNameStringProtocol.java
package com.tradeshift.reaktive.xml;

import javax.xml.namespace.QName;
import javax.xml.stream.events.XMLEvent;

import com.tradeshift.reaktive.marshal.Protocol;
import com.tradeshift.reaktive.marshal.TStringProtocol;

import io.vavr.Tuple2;

/**
 * Protocol for a tuple of QName and String, where the String can be transformed to other types
 * by invoking .as(TYPE).
 */
public class QNameStringProtocol extends TStringProtocol<XMLEvent,QName> {

    public QNameStringProtocol(Protocol<XMLEvent,Tuple2<QName, String>> delegate) {
        super(delegate, XMLProtocol.locator);
    }
    
    /**
     * Returns the protocol only reading and writing local names for _1, without namespace.
     */
    public TStringProtocol<XMLEvent,String> asLocalName() {
        return new TStringProtocol<>(this.map(t -> t.map1(QName::getLocalPart), t -> t.map1(QName::new)), XMLProtocol.locator);
    }
}

File: ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/QNameStringProtocol.scala
package com.tradeshift.reaktive.marshal.scaladsl

import com.tradeshift.reaktive.marshal.Protocol
import javax.xml.namespace.QName
import javax.xml.stream.events.XMLEvent
import com.tradeshift.reaktive.xml.XMLProtocol

/**
 * Protocol for a tuple of QName and String, where the String can be transformed to other types
 * by invoking .as(TYPE).
 */
class QNameStringProtocol(delegate: Protocol[XMLEvent,(QName, String)]) 
extends TStringProtocol[XMLEvent,QName](delegate, XMLProtocol.locator) {
  /**
   * Returns the protocol only reading and writing local names for _1, without namespace.
   */
  def asLocalName: TStringProtocol[XMLEvent,String] = new TStringProtocol(
    this.map(
      f1(t => t.copy(_1 = t._1.getLocalPart)),
      f1(t => t.copy(_1 = new QName(t._1)))
    ), XMLProtocol.locator)  
}